### PR TITLE
Add read interface for retrieving Vulnerability Remediation data

### PR DESCRIPTION
**Pull request by GabrielEValenzuela**
|Related issue|
|---|
|#19043|

## Objective

This pull request introduces a read interface that enables the retrieval of `Vulnerability Remediation` data from a RocksDB database. With this enhancement, users can access and query data related to security vulnerabilities and their remediation efforts more efficiently and reliably. The addition of this read interface significantly improves the system's capabilities, allowing for seamless integration of RocksDB as a data source, enhancing data retrieval speed, and providing a more robust foundation for managing security vulnerabilities within the application. This enhancement represents a critical step towards optimizing the performance and reliability of the Vulnerability Remediation feature.

## Description
- Implemented validation logic to verify the structure and content of the retrieved object.
- Added error-handling mechanisms to handle invalid or incomplete configurations.
- Included unit tests to verify the accuracy of the validation process.

For a better compression of the detailed work, please check this top-level sequence diagram

<details><summary>Sequence Diagram</summary>

image

</details>

## Quality Assurance
A new test has been included to verify the correct behavior and backward compatibility of this solution proposal.

<details><summary>Output testing 🟢 </summary>

```bash
[==========] Running 3 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 3 tests from DatabaseFeedManagerTest
[ RUN      ] DatabaseFeedManagerTest.GetVulnerabilityRemediation_ValidData
[       OK ] DatabaseFeedManagerTest.GetVulnerabilityRemediation_ValidData (72 ms)
[ RUN      ] DatabaseFeedManagerTest.GetVulnerabilityRemediation_DataNotFound
Key not found: CVE-2023-5678
[       OK ] DatabaseFeedManagerTest.GetVulnerabilityRemediation_DataNotFound (69 ms)
[ RUN      ] DatabaseFeedManagerTest.GetVulnerabilityRemediation_InvalidData
[       OK ] DatabaseFeedManagerTest.GetVulnerabilityRemediation_InvalidData (87 ms)
[----------] 3 tests from DatabaseFeedManagerTest (229 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test suite ran. (229 ms total)
[  PASSED  ] 3 tests
```

</details>

Regarding memory, no new memory allocations were introduced in this PR using `raw_pointers`.

<details><summary>ScanBuild report</summary>

Command:

```bash
scan-build-15 -enable-checker alpha.core.CastSize -enable-checker alpha.core.CastToStruct -enable-checker alpha.core.Conversion -enable-checker alpha.core.IdenticalExpr -enable-checker alpha.core.SizeofPtr -enable-checker alpha.core.TestAfterDivZero --exclude external make TARGET=server DEBUG=1 -j$(nproc)
```

```bash

Done building server

scan-build: No bugs found.

```

</details>

<details><summary>Clang-Format</summary>

```bash
---------------------------------
PASSED: Clang-format check passed
---------------------------------
```
</details>

## Exploratory tests
Not apply

### Check summary

- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Source installation
- [x] Package installation
- [x] Source upgrade
- [x] Package upgrade
- [ ] Review logs syntax and correct language
- [ ] QA templates contemplate the added capabilities

- Memory tests for Linux
  - [x] Scan-build report
  - [ ] Coverity
  - [ ] Valgrind (memcheck and descriptor leaks check)
  - [ ] AddressSanitizer

- [ ] Retrocompatibility with older Wazuh versions
- [ ] Working on cluster environments
- [ ] Configuration on demand reports new parameters
- [ ] The data flow works as expected (agent-manager-api-app)
- [x] Added unit tests (for new features)
- [ ] Stress test for affected components

### CODE REVIEW

**Review by MiguelazoDS — COMMENTED** [review]
The behavioral test is ok GJ!, although I have some concerns about including functions outside the test files. 

This for example

```c++
    // Decoy method call
    spDatabaseFeedManager->saveVulnerabilityRemediation();
```

It would be better to include this in a setup method for each test that requires it.

**Comment by GabrielEValenzuela** [comment]
I think is not possible, if we move it to the setup we need to move the database instance too and the test fails after that

File: src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp [hunk 1]
```diff
@@ -95,6 +128,71 @@ class TDatabaseFeedManager final : public Observer<nlohmann::json&>
...
+     */
+    void getVulnerabilityRemediation(std::string_view cveId, FlatbufferDataPair<RemediationInfo>& dtoVulnRemediation)
+    {
+        std::string key_str {cveId};
+
+        auto result = m_remediationsDatabase->get(key_str, dtoVulnRemediation.slice);
+
+        if (!result)
+        {
+            throw std::logic_error("Value for key " + key_str + " does not exist in database.");
```
Comment (matias-braida) [inline src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp:155]: I think these exceptions should be std::runtime. Based on the following explanation.

>     The class logic_error defines the type of objects thrown as exceptions to report errors presumably detectable before the program executes, such as violations of logical preconditions or class invariants.
> 
>     The class runtime_error defines the type of objects thrown as exceptions to report errors presumably detectable only when the program executes.
>

File: src/wazuh_modules/vulnerability_scanner/tests/databaseFeedManager/databaseFeedManager_test.cpp [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+    spPolicyManagerMock = std::make_shared<MockPolicyManager>();
+    EXPECT_CALL(*spPolicyManagerMock, getUpdaterConfiguration()).WillRepeatedly(Return(configurationParameters));
+
+    spContentRegisterMock = std::make_shared<MockContentRegister>(
+        configurationParameters.at("topicName").get<const std::string>(), configurationParameters);
+
+    spRouterSubscriberMock = std::make_shared<MockRouterSubscriber>(
+        configurationParameters.at("topicName").get<const std::string>(), "vulnerability_feed_manager");
+
```
Comment (matias-braida) [inline src/wazuh_modules/vulnerability_scanner/tests/databaseFeedManager/databaseFeedManager_test.cpp:63]: Using the following EXPECT_CALL you can remove the warning associated with this mock object.
`EXPECT_CALL(*spRouterSubscriberMock, subscribe(_));`

**Review by matias-braida — COMMENTED** [review]
Great Job

File: src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp (hunk 1 above)
Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp]: Fixed!

File: src/wazuh_modules/vulnerability_scanner/tests/databaseFeedManager/databaseFeedManager_test.cpp (hunk 2 above)
Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/tests/databaseFeedManager/databaseFeedManager_test.cpp:65]: Fixed. Thanks!

**Review by matias-braida — APPROVED** [review]
LGTM

**Review by MiguelazoDS — APPROVED** [review]

**Comment by GabrielEValenzuela** [comment]
Update
-----------

Testing after rebase 🟢 

```bash
[==========] Running 9 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 9 tests from DatabaseFeedManagerTest
[ RUN      ] DatabaseFeedManagerTest.GetVulnerabilityCandidatesSuccess
/home/gvalenzuela/Documents/Work/WazuhRepositories/vulnerability-refactor/src/wazuh_modules/vulnerability_scanner/schemas/vulnerabilityCandidate.fbs
[       OK ] DatabaseFeedManagerTest.GetVulnerabilityCandidatesSuccess (123 ms)
[ RUN      ] DatabaseFeedManagerTest.GetVulnerabilityCandidatesContextEmpty
/home/gvalenzuela/Documents/Work/WazuhRepositories/vulnerability-refactor/src/wazuh_modules/vulnerability_scanner/schemas/vulnerabilityCandidate.fbs
[       OK ] DatabaseFeedManagerTest.GetVulnerabilityCandidatesContextEmpty (124 ms)
[ RUN      ] DatabaseFeedManagerTest.GetVulnerabilityCandidatesNoPackageName
/home/gvalenzuela/Documents/Work/WazuhRepositories/vulnerability-refactor/src/wazuh_modules/vulnerability_scanner/schemas/vulnerabilityCandidate.fbs
Invalid package
[       OK ] DatabaseFeedManagerTest.GetVulnerabilityCandidatesNoPackageName (144 ms)
[ RUN      ] DatabaseFeedManagerTest.GetVulnerabilityRemediation_ValidData
/home/gvalenzuela/Documents/Work/WazuhRepositories/vulnerability-refactor/src/wazuh_modules/vulnerability_scanner/schemas/vulnerabilityCandidate.fbs
[       OK ] DatabaseFeedManagerTest.GetVulnerabilityRemediation_ValidData (115 ms)
[ RUN      ] DatabaseFeedManagerTest.GetVulnerabilityRemediation_DataNotFound
/home/gvalenzuela/Documents/Work/WazuhRepositories/vulnerability-refactor/src/wazuh_modules/vulnerability_scanner/schemas/vulnerabilityCandidate.fbs
Key not found: CVE-2023-5678
[       OK ] DatabaseFeedManagerTest.GetVulnerabilityRemediation_DataNotFound (96 ms)
[ RUN      ] DatabaseFeedManagerTest.GetVulnerabilityRemediation_InvalidData
/home/gvalenzuela/Documents/Work/WazuhRepositories/vulnerability-refactor/src/wazuh_modules/vulnerability_scanner/schemas/vulnerabilityCandidate.fbs
[       OK ] DatabaseFeedManagerTest.GetVulnerabilityRemediation_InvalidData (115 ms)
[ RUN      ] DatabaseFeedManagerTest.getVulnerabiltyDescriptiveInformation_Ok
[       OK ] DatabaseFeedManagerTest.getVulnerabiltyDescriptiveInformation_Ok (126 ms)
[ RUN      ] DatabaseFeedManagerTest.getVulnerabiltyDescriptiveInformation_NotFound
/home/gvalenzuela/Documents/Work/WazuhRepositories/vulnerability-refactor/src/wazuh_modules/vulnerability_scanner/schemas/vulnerabilityCandidate.fbs
Key not found: cveid_any
[       OK ] DatabaseFeedManagerTest.getVulnerabiltyDescriptiveInformation_NotFound (119 ms)
[ RUN      ] DatabaseFeedManagerTest.getVulnerabiltyDescriptiveInformation_Corrupted
/home/gvalenzuela/Documents/Work/WazuhRepositories/vulnerability-refactor/src/wazuh_modules/vulnerability_scanner/schemas/vulnerabilityCandidate.fbs
[       OK ] DatabaseFeedManagerTest.getVulnerabiltyDescriptiveInformation_Corrupted (121 ms)
[----------] 9 tests from DatabaseFeedManagerTest (1088 ms total)

[----------] Global test environment tear-down
[==========] 9 tests from 1 test suite ran. (1088 ms total)
[  PASSED  ] 9 tests.
```

**Pull request merged**